src/cephadm/cephadm.py

```diff
@@ -5783,6 +5792,43 @@ def copy_file(src: str, dst: str) -> None:
         logger.warning(f'Cannot create cluster configuration directory {conf_dir}')
 
 
+def rollback(func: FuncT) -> FuncT:
+    """
+    """
+    @wraps(func)
+    def _rollback(ctx: CephadmContext) -> Any:
+        try:
+            return func(ctx)
+        except ClusterAlreadyExists:
+            # another cluster with the provided fsid already exists: don't remove.
+            raise
+        except (KeyboardInterrupt, Exception) as e:
+            logger.error(f'{type(e).__name__}: {e}')
+            if not ctx.keep_cluster_files:
+                logger.info('\n\n'
+                            '\t***************\n'
+                            '\tcannot proceed with ceph cluster installation. Current cluster files will be deleted automatically,\n'
+                            '\tto disable this behaviour you can pass the --keep-cluster-files flag. In case of any previous\n'
+                            '\tbroken installation user must use the following command to completely delete the broken cluster:\n\n'
+                            '\t> cephadm rm-cluster --force --zap-osds --fsid <fsid>\n\n'
+                            '\tfor more information please refer to https://docs.ceph.com/en/latest/cephadm/operations/#purging-a-cluster\n'
+                            '\t***************\n\n')
+                _command_rm_cluster(ctx, keep_logs=False, zap_osds=False)
+            else:
+                logger.info('\n\n'
+                            '\t***************\n'
+                            '\tcannot proceed with ceph cluster installation. Current cluster files will NOT BE DELETED automatically to change\n'
+                            '\tthis behaviour you can pass the --no-keep-cluster-files flag. To remove this broken cluster manually please run:\n\n'
+                            f'\t   > cephadm rm-cluster --force --fsid {ctx.fsid}\n\n'
+                            '\tin case of any previous broken installation user must use the rm-cluster command to delete the broken cluster:\n\n'
+                            '\t   > cephadm rm-cluster --force --zap-osds --fsid <fsid>\n\n'
+                            '\tfor more information please refer to https://docs.ceph.com/en/latest/cephadm/operations/#purging-a-cluster\n'
+                            '\t***************\n\n')
+            raise
```

Comment:
is this `raise` reachable? The second `except` is catching the generic `Exception` type, so I don't see how anything would get past it.

Reply:
this raise role is to propagate the exception so the block below can capture the exception, print the error and `sys.exit(1)` ... it can be replaced by a `sys.exit(1)` but I preferred to maintain previous exception handling in the same place. This way, error handling is done in one point only. In case you feel the `raise` is hidden I can add some comment.

```
    except (Error, ClusterAlreadyExists) as e:
        if ctx.verbose:
            raise
        logger.error('ERROR: %s' % e)
        sys.exit(1)

```

Reply:
sorry, I misread the indentation of this raise when reviewing last time. This makes total sense.